DM-32442: Add Java unit tests for SAL.

wvreeven opened the pull request.

dmills-lsst reviewed (commented):
So far so good...
BaseTestCase.java references SALPY in comment
LsstDdsQosTest,java has /home/wreeven path
Does @throws Exception In case of a problem related to reflection mean the tests do not work?

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java:
```diff
@@ -0,0 +1,31 @@
+
+/**
+ * A few tests require CSC that doesn't use generics, so we use Script.
+ */
```
The way SAL handles generic topics has changed, making this comment and the test names and details somewhat out of date. The XML field name has changed from Generics to AddedGenerics and "no" is no longer a valid entry. But this is fundamentally a test of the way that SAL handles generic topics, and Script is still unusual in not being a CSC.

Please at least update the code comment, test class name and test name(s) to reflect the current reality.

Please also consider updating the test itself, e.g.:
* Test that `Script` has no `summaryState` event nor `enable` command (two topics that all CSCs have). This is a useful change from testing for the `enterControl` command, because now very CSCs have the `enterControl` command so it is normal for it to be missing.
* Test that `Test` has no `enterControl` command.

r-owen replied:
If you prefer, I can update the Python code first.

wvreeven replied:
Please update the Python code first. Unfortunately I am very busy with the dome software installation at the summit and therefore prefer to not have to spend too much time on this ticket if I can possibly help it.

r-owen replied:
I filed DM-32956 to fix the Python and implemented it in https://github.com/lsst-ts/ts_sal/pull/168.

I would rather have this bit of Java cleaned up before merging, even if it adds a delay while you deal with more important things. I am willing to do the cleanup myself, if need be.

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:
```diff
@@ -0,0 +1,427 @@
+    @Test
+    @Ignore("No astropy equivalent for java.")
+    public void testGetLeapSeconds() {
+    }
```
Surely Dave is willing to add getLeapSeconds to Java?

I'd be a lot more comfortable if these issues you are finding in the unit tests get fixed in SAL. Dave usually fixes things quickly and if anything he can fix quickly would be a big win over committing disabled tests.

wvreeven replied:
The way Java handles leap seconds probably is such that this cannot be fixed by Dave. Java basically makes a smooth transition by enlarging the time span of a second during a certain period of time so no "additional" second is needed. It is very ugly and bad, but that's the way it is. If you prefer I can remove the test case.

r-owen replied:
This seems very surprising. The value returned by SAL's time function (which is not this one) is the value of CLOCK_TAI. I don't think Java would change that. I don't know how ts_sal computes the value returned by getLeapSeconds, but again it should be easy to get that value through to Java. In any case I've started a discussion on Slack to learn more about this and we can revisit this based on that discussion.

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:
```diff
@@ -0,0 +1,427 @@
+     * @throws Exception In case of a problem related to reflection.
+     */
+    @Test
+    @Ignore("Fails due to JNI limitations if ran with all other unit tests.")
```
"ran" -> "run"

This is getting really scary. This is a very important test for serious problems we have had. Please try to find some way to kill the JVM between test invocations.

Perhaps using a different LSST_DDS_DOMAIN_PREFIX for each test invocation will fix this particular issue?

wvreeven replied:
See my previous remark about forking the JVM when running the unit tests. This issue will be addressed in DM-32914.

r-owen commented on java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Before
+    public void setUp() {
+        Random rand = new Random(2000L);
```
Please generate a random LSST_DDS_DOMAIN_PREFIX env var before constructing the remote and controller. That is potentially useful for keeping tests from colliding with each other. I think it won't be a problem that the JVM caches information, because SAL reads the env var when it creates the `SAL_<name>` object.

wvreeven replied:
Would it be OK if I addressed this as well in DM-32914?

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:
```diff
@@ -0,0 +1,427 @@
+    /**
+     * Write several logevent messages and make sure gettting the oldest returns the data in the expected order.
+     *
+     * @throws Exception In case of a problem related to reflection.
```
What does `reflection` mean in this context? Please expand this comment (and the many others like it).

wvreeven replied:
Reflection is a Java standard present in the JDK and JVM for introspection of classes and class instances. Much like getattr in Python. Any Java developer knows what reflection means in a Java context so from a Java point of view there is no need to explain this.

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:
```diff
@@ -0,0 +1,427 @@
+     * Only one value is retrievable.
+     */
+    @Test
+    @Ignore("Doesn't work because of JNI library loading and garbage collection.")
```
The ability to read late-joiner events is absolutely critical. It feels like a very large and important swath of SAL is not tested due to this JVM issue.

wvreeven replied:
See my previous remark about forking the JVM when running the unit tests. This issue will be addressed in DM-32914.

r-owen commented on java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:
```diff
@@ -0,0 +1,427 @@
+    }
+
+    /**
+     * Test that a late joiner can an event using getNextSample.
```
"can an event" -> "can read the most recent event" or similar.

wvreeven replied:
Again, I implemented the python tests as literal as possible meaning that I copied the python doc strings as well. I've fixed this in both the python and the Java files.

r-owen reviewed (changes requested):
This is a great start, but with some worrying limitations:
* Some important tests are disabled due to a JVM issue. I think it is important to try to figure out how to run these (though I won't block this ticket on that basis as long as you file a ticket to fix this problem).
* Other tests are disabled because SAL itself is apparently misbehaving. For these I think it is important to ask Dave to fix the problems, and if he can do it quickly, then I would strongly prefer to hold off merging this until the problems are fixed and the tests can be enabled.

It is possible that a few of the JVM-related issues can be worked around by setting a random LSST_DDS_DOMAIN_PREFIX before constructing each pair of SAL objects. It is well worth trying, especially as it prevents interference with production code. (If you are doing that, I missed it)

wvreeven commented on doc/version_history.rst:
```diff
@@ -44,6 +44,8 @@ Changes for 6.0.0
 
 * Change getXXXVersion's to static for C++
 
+* Add Java unit tests for SAL
```
Good point! Not sure why I didn't realize that myself. Fixed.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:
```diff
@@ -0,0 +1,75 @@
+    @Before
+    public void setUp() {
+        Random rand = new Random(2000L);
+        int index = rand.nextInt();
```
I made sure that it always is positive.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:
```diff
@@ -0,0 +1,75 @@
+        final Method m = cls.getMethod(funcName, data.getClass());
+        final long startTime = System.currentTimeMillis();
+        Integer retcode = SAL_Test.SAL__NO_UPDATES;
+        while (System.currentTimeMillis() - startTime < 2000L) {
```
Done.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+        try {
+            controller.salEventPub(badEvtName);
+            // This Assert should work but doesn't.
+            // Assert.fail("An exception should have been thrown.");
```
I think it is yes. I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

wvreeven replied:
Fixed.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+
+        // These next lines don't work in Java due to strong typing.
+        // command_setArrays bad_data = new command_setArrays();
+        // checkCmdGetPutRaises(bad_data);
```
I added them to make sure that it was clear that the python tests were implemented as literally as possible. That goal was achieved and I have removed these lines.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+        // checkEvtGetPutRaises(bad_data);
+
+        // This line doesn't work in Java. No exception is thrown.
+        // checkEvtGetPutRaises(null);
```
I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

wvreeven replied:
I have opened DM-32952 for this.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+
+        // These next lines don't work in Java due to strong typing.
+        // setArrays bad_data = new setArrays();
+        // checkTelGetPutRaises(bad_data);
```
Done

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+        // checkTelGetPutRaises(bad_data);
+
+        // This line doesn't work in Java. No exception is thrown.
+        // checkTelGetPutRaises(null);
```
I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

wvreeven replied:
I have opened DM-32952 for this.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+    @Test
+    public void testTelOverflowBuffer() throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+        remote.salTelemetrySub("Test_scalars");
+        sleep(STD_SLEEP);
```
I have literally implemented the Python tests and have not checked if the sleeps are necessary or not. OK if I move checking that and removing the unnecessary sleeps in another ticket?

wvreeven replied:
I have opened DM-32952 for this.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+     * Check that sending too long a string causes an error.
+     */
+    @Test
+    @Ignore("This test actually crashes the JVM.")
```
I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

wvreeven replied:
I have opened DM-32952 for this.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+     * Check that setting sending too long an array causes an error.
+     */
+    @Test
+    @Ignore("Java works in a way that this unit test will never pass.")
```
Done

wvreeven commented on java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:
```diff
@@ -0,0 +1,388 @@
+        }
+    }
+
+    private void checkCmdGetPutRaises(command_setScalars data) {
```
Java has MANY checked exceptions that require the developer to either surround a statement with a try/catch block or declare that the calling method throws the exception. The methods that get called in the methods declare to throw checked exceptions, which is why they need to be either caught or re-thrown. The pattern applied here is the accepted pattern for JUnit to make sure that an exception is thrown by the method being called while still being able to continue in the method. I've added a comment briefly explaining that but any Java developer is supposed to know how to handle checked exceptions.

wvreeven commented on java_tests/src/test/java/org/lsst/sal/LsstDdsQosTest.java:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * This unit test class doesn't make much sense in Java due to the way JNI loads the
+ * native libraries and doesn't unload them until the JVM stops.
```
Thanks for this comment. It made me think and it actually turns out that that is possible. The JVM in which the unit tests are run can be forked such that that actual unit test code itself runs in a separate JVM. And it also is possible to make sure that a forked JVM is not reused, meaning that each unit test class starts with a fresh JVM. I opened DM-32914 for this.

wvreeven commented:
Thanks for all the useful comments. Regarding the question raised by Dave "Does @throws Exception In case of a problem related to reflection mean the tests do not work?", the answer is no. Java has the concept of checked exceptions which mean that you either need to try/catch the exception or declare that your method throws the exception for the compiler to be happy. I chose the latter since we know that the class members, for which the values are being set, exist and are accessible (meaning they are public and not private).

r-owen reviewed (changes requested):
Thanks for filing the new tickets. A few things remain:
* I think the version number needs a smaller jump.
* I would still like to see the AddedGenerics test updated as per my suggestion. I'm willing to do this if you don't have time. But I think it would be very quick. I updated the python code on https://github.com/lsst-ts/ts_sal/pull/168 if you want an example (though I cannot yet it yet because I can't build the Script SALPY library; I've filed a ticket about that).
* I am not yet convinced by the argument about getLeapSeconds, though that can be postponed to one of the other tickets.

r-owen reviewed (approved):
With @wvreeven 's kind permission I updated the generics test for modern AddedGenerics tag (matching a very recent update to the SALPY unit tests) and rebased the PR.

I also changed the version number to 6.1.0.

Please run the tests to make sure I didn't break anything. Feel free to merge once you are happy, and push back if you don't like my changes.

One more suggestion to @wvreeven and @dmills-lsst: please consider not releasing ts_sal 6.1 until the related tickets are done, so that the Java issues you discovered (that required disabling tests) are fixed, and the Java tests are running (to keep these issues fixed). There's no rush and new SAL versions tend to be a big deal.

wvreeven commented on doc/version_history.rst:
```diff
@@ -9,18 +9,23 @@ Version History
+Changes for 7.0.0
+=================
+
+* Add Java unit tests for SAL
```
I lowered it to 6.1.0

wvreeven commented:
I adopted the doc strings for the checkXGetPutRaises in the Java tests code.
I adopted the python changes to AddedGenerics in the SAL_ScriptTest.java code.
I added a crude getLeapSecondsTest which verifies that the returned value for leap seconds is greater than or equal to zero.

Also, I modified the next version for sal from 6.0.1 (as added by Russell) to 6.1.0.
